fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewPushDownRewriter.java

```diff
@@ -0,0 +1,857 @@
+                    ScalarOperator rewritten = rewriter.rewrite(entry.getValue());
+                    aggProjection.put(entry.getKey(), rewritten);
+                }
+            } else {
+                // If there is no projections before, aggregations are already in the aggProjection.
+                // group by keys
+                for (ColumnRefOperator columnRefOperator : aggregate.getGroupingKeys()) {
+                    aggProjection.put(columnRefOperator, columnRefOperator);
+                }
+            }
+
+            // rewrite aggregate's predicate
+            ScalarOperator predicate = aggregate.getPredicate();
+            if (aggregate.getPredicate() != null) {
+                predicate = rewriter.rewrite(aggregate.getPredicate());
+            }
+
+            Projection projection = new Projection(aggProjection);
+            LogicalAggregationOperator newAgg = LogicalAggregationOperator.builder()
+                    .withOperator(aggregate)
+                    .setAggregations(newAggregations)
+                    .setProjection(projection)
+                    .setPredicate(predicate)
+                    .build();
+            optExpression = OptExpression.create(newAgg, optExpression.getInputs());
+            if (!checkAggOpt(optExpression)) {
+                return AggRewriteInfo.NOT_REWRITE;
+            }
+            rewriteInfo.setOp(optExpression);
+            return rewriteInfo;
+        }
+
+        // rewrite it with remapping and final aggregate should use the new input as its argument.
+        private CallOperator replaceAggFuncArgument(Map<ColumnRefOperator, ColumnRefOperator> remapping,
+                                                    ColumnRefOperator origAggColRef,
+                                                    CallOperator aggCall,
+                                                    int argIdx) {
+            ScalarOperator newArg0 = remapping.get(origAggColRef);
+            if (newArg0 == null) {
+                logMVRewrite(mvRewriteContext, "Aggregation's arg0 is not rewritten after remapping, " +
+                        "aggColRef:{}, aggCall:{}", origAggColRef, aggCall);
+                return null;
+            }
+            CallOperator newAggCall = (CallOperator) aggCall.clone();
+            if (argIdx >= newAggCall.getChildren().size()) {
+                logMVRewrite(mvRewriteContext, "Aggregation's arg index is out of range, " +
+                        "aggColRef:{}, aggCall:{}", origAggColRef, aggCall);
+                return null;
+            }
+            newAggCall.setChild(argIdx, newArg0);
+            return newAggCall;
+        }
+
+        @Override
+        public AggRewriteInfo visitLogicalJoin(OptExpression optExpression,
+                                               AggRewriteInfo rewriteInfo) {
+            AggregatePushDownContext context = rewriteInfo.getCtx();
```

Comment:
Do we support the pattern?  and we push the agg down to join3?
agg
|
join1
|      \
join2   join3

Reply:
No, this pr can only support agg push down to scan node, and agg->scan can be rewritten by `LogicalSPG` MV

Reply:
I think agg - join can also be supported later.